File: src/jsonrpc/eth/getTransactionByHash/eth_getTransactionByHash.js
/**
 * @fileoverview eth_getTransactionByHash JSON-RPC method
 */

import { createRequest } from "../../types/JsonRpcRequest.js";

/**
 * @typedef {import('../../types/index.js').Address} Address
 * @typedef {import('../../types/index.js').Hash} Hash
 * @typedef {import('../../types/index.js').Quantity} Quantity
 * @typedef {import('../../types/index.js').BlockTag} BlockTag
 * @typedef {import('../../types/index.js').BlockSpec} BlockSpec
 * @typedef {import('../../types/JsonRpcRequest.js').JsonRpcRequest} JsonRpcRequest
 */

/**
 * Returns the information about a transaction requested by transaction hash.
 *
 * @example
 * Transaction hash: "0xa52be92809541220ee0aaaede6047d9a6c5d0cd96a517c854d944ee70a0ebb44"
 * Result: ...
 *
 * Implements the `eth_getTransactionByHash` JSON-RPC method.
 */

/** The JSON-RPC method name */
export const method = "eth_getTransactionByHash";
/**
 * Result for `eth_getTransactionByHash`
 *
 * @typedef {Quantity} Result
 */

/**
 * Request for `eth_getTransactionByHash`
 *
 * @typedef {JsonRpcRequest<'eth_getTransactionByHash', [Hash]>} Request
 */

/**
 * Creates a eth_getTransactionByHash JSON-RPC request
 *
 * @param {Hash} address
 * @param {number | string | null} [id] - Optional request ID
 * @returns {Request}
 */
export function GetTransactionByHashRequest(address, id = null) {
	return /** @type {Request} */ (createRequest(method, [address], id));
}

File: src/jsonrpc/eth/getTransactionByHash/eth_getTransactionByHash.zig
const std = @import("std");
const types = @import("../../types.zig");

/// Returns the information about a transaction requested by transaction hash.
///
/// Example:
/// Transaction hash: "0xa52be92809541220ee0aaaede6047d9a6c5d0cd96a517c854d944ee70a0ebb44"
/// Result: ...
///
/// Implements the `eth_getTransactionByHash` JSON-RPC method.
pub const EthGetTransactionByHash = @This();

/// The JSON-RPC method name
pub const method = "eth_getTransactionByHash";

/// Parameters for `eth_getTransactionByHash`
pub const Params = struct {
    /// 32 byte hex value
    transaction hash: types.Hash,

    pub fn jsonStringify(self: Params, jws: *std.json.Stringify) !void {
        try jws.beginArray();
        try jws.write(self.transaction hash);
        try jws.endArray();
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Params {
        if (source != .array) return error.UnexpectedToken;
        if (source.array.items.len != 1) return error.InvalidParamCount;

        return Params{
            .transaction hash = try std.json.innerParseFromValue(types.Hash, allocator, source.array.items[0], options),
        };
    }
};

/// Result for `eth_getTransactionByHash`
pub const Result = struct {
    value: types.Quantity,

    pub fn jsonStringify(self: Result, jws: *std.json.Stringify) !void {
        try jws.write(self.value);
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Result {
        return Result{
            .value = try std.json.innerParseFromValue(types.Quantity, allocator, source, options),
        };
    }
};
